### ISSUE TITLE
Yum module's install action does not verify if the installed package's architecture is the same when it decides if the package has already been installed

### ISSUE THREAD

**Issue by vlajos** [ansible/ansible#66894]
##### SUMMARY
In this section:
https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/packaging/os/yum.py#L977
Line 989 is triggered regardless if the already installed package is from a different architecture.

That way you cannot install a package's i686 version if the the same version's x86_64 has already been installed previously.

- Bug Report

##### COMPONENT NAME
lib/ansible/modules/packaging/os/yum.py

##### ANSIBLE VERSION
I have checked since 2.7.4

##### CONFIGURATION
I do not think it is config specific.

##### OS / ENVIRONMENT
I do not think it is OS specific.

##### STEPS TO REPRODUCE

Install any installed x86_64 package's i686 version.

For example:
```
rpm -qa|grep glibc
glibc-2.17-292.el7.x86_64
```

```
- name: Install another arch's package
  yum:
    name:  glibc-2.17-292.el7.i686
    state: present
```

##### EXPECTED RESULTS

The package should be installed.

##### ACTUAL RESULTS

The package does not get installed.

**Comment by ansibot** [comment]
Files identified in the description:
* [`lib/ansible/modules/packaging/os/installp.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/packaging/os/installp.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @MorrisA @bcoca @d-little @dagwieers @flynn1973 @gforster @kairoaraujo @marvin-sinister @mator @molekuul @ramooncamacho @wtcross
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
Files identified in the description:
* [`lib/ansible/modules/packaging/os/yum.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/packaging/os/yum.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @Akasurde @berenddeschouwer @kustodian @maxamillion @verm666
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by drgrigorov** [comment]
Hello, I have also hit this same behavior with ansible 2.8.3.
I have spent some time debugging it and found that yum module does not account for arch at all when it finds installed rpm package (while using direct rpm):

```python
 969                     if compare > 0 and self.allow_downgrade:
 970                         downgrade_candidate = True
 971                     elif compare >= 0:
 972                         continue
```
I have changed to "elif compare >= 0 and cur_arch == arch:" which resulted in actual success. I am not sure if this patch is safe or robust, just pointing my findings.

**Comment by ansibot** [comment]
Files identified in the description:
* [`lib/ansible/modules/yum.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/yum.py)

If these files are incorrect, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by mkrizek** [comment]
I was not able to reproduce the issue on centos 7 docker container with `glibc-2.17-317.el7.x86_64` installed and the following playbook:
```yaml
- hosts: centos7
  gather_facts: false
  tasks:
    - name: Install another arch's package
      yum:
        name: glibc-2.17-317.el7.i686
        state: present
```

needs_info

**Comment by drgrigorov** [comment]
> I was not able to reproduce the issue on centos 7 docker container with `glibc-2.17-317.el7.x86_64` installed and the following playbook:
> 
> ```yaml
> - hosts: centos7
>   gather_facts: false
>   tasks:
>     - name: Install another arch's package
>       yum:
>         name: glibc-2.17-317.el7.i686
>         state: present
> ```
> 
> needs_info

Hi mkrizek, which ansible version did you try? As of 2.9.6 I don't see any change in code in that module specifically for the snippet I mentioned. What is the currently installed version of glibc in your container and what is the result of the playbook you tried? Let me try to reproduce meanwhile

**Comment by mkrizek** [comment]
Tested on all supported version: `devel`, `2.11.4`, `2.10.13` and `2.9.25`.

**Comment by drgrigorov** [comment]
I could not reproduce it with:
ansible 2.9.6 + python 3.8.10 with target centos7 (docker). It is either fixed or we were hitting some edge case. I don't remember the package I were attempting to install while this was found.

**Comment by mkrizek** [comment]
Closing per above. If the issue persists, please open a new issue with a simple reproducer including a package to test against.

If you have further questions please stop by IRC or the mailing list:
* IRC: #ansible on irc.libera.chat
* mailing list: https://groups.google.com/forum/#!forum/ansible-project

**Comment by vlajos** [comment]
Hi,
I have just checked this again. (Sorry for the late response...)
I think without `use_backend: yum` the yum module uses the `dnf` backend which is probably bug free.
I tried to create an example, but it seems that the module is not compatible with python3:
> fatal: [localhost]: FAILED! => {"changed": false, "msg": "The Python 2 yum module is needed for this module. If you require Python 3 support use the `dnf` Ansible module instead."}

At least I could not make it working with python2.

I guess it is not really worth spending more time on debugging something depreacated.

Thanks,
Lajos
